Developing-new-features-with-feature-gating.md

```diff
@@ -93,9 +132,37 @@ We use the following principles to determine the value of a feature flags:
+
+## How do you, as a developer, use feature flags?
+
+### When to use feature flags?
+
+Feature flags should be used when you are working on a feature whose scale makes it hard to implement it all at once and will need to be added incrementally/if the feature is likely to cause breakages. Essentially, feature flags are a must for featues that are not ready for production/fully-tested yet.
+
+### How to use feature flags?
+
+Say you are working on a large scale user-facing feature that will take multiple PRs to be fully implemented. The following is the recommended way to use feature flags in such a scenario (*recommended implies that this is the way we use/have used feature flags in the past in Oppia. We haven't encountered a scenario where we required an alternate workflow. Unless there are compelling reasons, this is the development workflow we expect developers to follow when it comes to employing the use of feature flags*):
+
+1. The very first PR you make must include changes that add a new feature flag to the codebase, meant to be used with this new feature. The feature flag must be placed in the DEV stage. Every single user-facing aspect of the feature -- whether frontend or backend -- must be gated behind the feature flag (both in the first PR, and all the following ones as well). This is to ensure that the feature is not visible to the user until it is fully implemented and ready for production.
+
+2. The first PR must be merged before any following PRs are merged. This is to ensure that the feature flag is available in the codebase for it to be used in the following PRs.
+
+3. The very last PR you make (to finish up the feature you are working on) must include changes that move the feature flag to the TEST stage. This is to ensure that the feature is available in the test environment, and we can feature-test it before it is made available to the users in the production environment.
+
+4. If the feature testing reveals that the feature is not ready for production, you must work on fixing the highlighted issues before proceeding further.
 
-For example, if we want to make a feature flag to be enable only in dev environments, we can set the feature flag as the image below.
+5. Once you receive a go ahead from the testers, you must merge another PR, moving the feature flag to the PROD stage, allowing it to be enabled/disabled in production (by the admin(s)).
 
-![Example feature setting](https://i.imgur.com/GnFQ7El.jpg)
\ No newline at end of file
+6. Once the feature is confirmed to be functioning as intended, you must merge one last PR to essentially "un-gate" the feature. Note that the flag itself will remain in the codebase (i.e. in the platform_parameter_list.py file, etc.), it just won't be used anymore (for example, all the `if` blocks and such that were used to gate the feature until this point must be removed).
```

Comment:
as intended --> as intended in production for at least 2 weeks

Shouldn't we also just get rid of the flag altogether at this point? I don't see why we need to keep it in the parameter list.

Reply:
Yep, modified this point

Reply:
> Shouldn't we also just get rid of the flag altogether at this point? I don't see why we need to keep it in the parameter list.

I think the flag, once added to the codebase and after making its way into the prod server, ends up in the datastore. Since flag names must be unique, the flag is moved to the deprecated list (I've updated the instructions) to allow for developers to see how to name new feature flags so there aren't any conflicts.

I have no source for this information (the feature flags feature isn't documented all that well), this is just my guess as to why we can't just remove the feature flags and instead have to move them into the deprecated list and have them stick around even after we're done using them.

Reply:
Ah, I see. Yeah I think you're right about them moving into the datastore. In step 6 can you also ask developers to also remove the flag entirely from the codebase except from that one part?